blockchain/chain.go

```diff
@@ -1290,6 +1290,176 @@ func (b *BlockChain) HeightRange(startHeight, endHeight int32) ([]chainhash.Hash
+	stopNode := b.index.LookupNode(hashStop)
+	if len(locator) == 0 {
+		// No blocks with the stop hash were found so there is nothing
+		// to do.
+		if stopNode == nil {
+			return nil
+		}
+		return []chainhash.Hash{*hashStop}
+	}
+
+	// Find the most recent locator block hash in the main chain.  In the
+	// case none of the hashes in the locator are in the main chain, fall
+	// back to the genesis block.
+	knownNode := b.bestChain.Genesis()
+	for _, hash := range locator {
+		node := b.index.LookupNode(hash)
+		if node != nil && b.bestChain.Contains(node) {
+			knownNode = node
+			break
+		}
+	}
+
+	// Start at the block after the most most recently known block.  When
+	// there is no next block it means the most recently known block is
+	// the tip of the best chain, so there is nothing more to do.
+	knownNode = b.bestChain.Next(knownNode)
+	if knownNode == nil {
+		return nil
+	}
+
+	// Calculate how many hashes will be returned.
+	total := uint32((b.bestChain.Tip().height - knownNode.height) + 1)
+	if stopNode != nil && b.bestChain.Contains(stopNode) &&
+		stopNode.height >= knownNode.height {
+
+		total = uint32((stopNode.height - knownNode.height) + 1)
+	}
+	if total > maxHashes {
+		total = maxHashes
+	}
+
+	// Populate and return the found hashes.
+	hashes := make([]chainhash.Hash, 0, total)
+	for node := knownNode; node != nil; node = b.bestChain.Next(node) {
+		hashes = append(hashes, node.hash)
+
+		// Stop once either the number of hashes exceeds the maximum or
+		// the stop hash has been reached.
+		if uint32(len(hashes)) >= maxHashes || node == stopNode {
+			break
+		}
+	}
+	return hashes
+}
+
+// LocateBlocks returns the hashes of the blocks after the first known block in
+// the locator until the provided stop hash is reached, or up to the provided
+// max number of block hashes.
+//
+// In addition, there are two special cases:
+//
+// - When no locators are provided, the stop hash is treated as a request for
+//   that block, so it will either return the stop hash itself if it is known,
+//   or nil if it is unknown
+// - When locators are provided, but none of them are known, hashes starting
+//   after the genesis block will be returned
+//
+// This function is safe for concurrent access.
+func (b *BlockChain) LocateBlocks(locator BlockLocator, hashStop *chainhash.Hash, maxHashes uint32) []chainhash.Hash {
+	b.chainLock.RLock()
+	hashes := b.locateBlocks(locator, hashStop, maxHashes)
+	b.chainLock.RUnlock()
+	return hashes
+}
+
+// locateHeaders returns the headers of the blocks after the first known block
+// in the locator until the provided stop hash is reached, or up to the provided
+// max number of block headers.
+//
+// See the comment on the exported function for more details on special cases.
+//
+// This function MUST be called with the chain state lock held (for reads).
+func (b *BlockChain) locateHeaders(locator BlockLocator, hashStop *chainhash.Hash, maxHeaders uint32) []wire.BlockHeader {
```

Comment:
I'm a bit torn on this one.  That is originally how I had it, but I decided to change it over as an optimization to avoid the temporary slice allocation.  As the commit description says `Since the majority of the network has moved to header-based semantics, this also provides an additional optimization to allow headers to be located directly versus needing to first discover the hashes and the fetch the headers.`

I can certainly change it if there are any serious reservations about the duplication, but my goal was really to optimize for the most common case, and given that realistically almost every peer on the network now uses headers-first semantics, it is the most common case.